### PR TITLE
Escape Attribute term

### PR DESCRIPTION
**Pull request by JesserH**
## Context

* We want to users to not be able of entering possible harmful javascript code into the schema using the attributes

## Summary

This PR can be summarized in the following changelog entry:

* Fixes a bug where possible harmful javascript could be entered in the database and affect a website.

## Relevant technical choices:

* Chosen for `WPSEO_Utils::format_json_encode` over `json_encode` because it is conform the Yoast standards
* Chosen to remove the first and last character of the string if they happen to be double quotes, because otherwise the schema output would include 4 double quotes in total (""example"")

## Test instructions

This PR can be tested by following these steps:

1. Install Yoast SEO (Premium), Woocommerce and wpseo woocommerce
2. Create a product attribute names `brand` and add a term called `banana`
3. In SEO -> WooCommerce SEO -> set Brand to `Product brand`
4. Create a product, give it a price and assign the brand attribute, set it to use `banana`. Save the product.
5. In your database, go to `<wp-prefix>_terms` and lookup `banana`. Change the name to `</script><script>alert(1)</script><script>`
6. Visit the product on the frontend, note that no pop-up shows
7. Inspect the source code and note that in the product graph the `name` of `"brand"` is ""
8. In your database, change the name to `</script><h1>hello</h1><script>`
9. Visit the product on the frontend, note that no `h1` shows and the additional information is the same as the name in the database 
10. Inspect the source code and note that in the product graph the `name` of `"brand"` is "hello"
11. In your database change the name to `Gimli`
12. Visit the product on the frontend, note that the additional information is the same as the name in the database (Gimli)
13. Inspect the source code and note that in the product graph `name` is "Gimli"
14. In your database change the name to `"Gimli"` (with the double quotes)
15. Visit the product on the frontend, note that the additional information is the same as the name in the database ("Gimli") with quotes
16. Inspect the source code and note that in the product graph the quotes are escaped

Fixes Yoast/bugreports#927

### CODE REVIEW

File: classes/woocommerce-schema.php [hunk 1]
```diff
@@ -300,12 +300,20 @@ private function add_manufacturer( $product ) {
...
+		$tag_stripped_term_name = wp_strip_all_tags( $escaped_term_name );
+		if ( $tag_stripped_term_name[0] === '"' && $tag_stripped_term_name[ ( strlen( $tag_stripped_term_name ) - 1 ) ] === '"' ) {
+			$tag_stripped_term_name = substr( $tag_stripped_term_name, 1, -1 );
+		}
 
 		if ( $term !== null ) {
 			$this->data[ $attribute ] = [
 				'@type' => 'Organization',
-				'name'  => $term->name,
+				'name'  => $tag_stripped_term_name,
```
Comment (karlijnbok) [inline classes/woocommerce-schema.php:316]: Maybe rename `$escaped_term_name` and `$tag_stripped_term_name` to `$sanitized_term_name`?
> $sanitized_term_name      = WPSEO_Utils::format_json_encode( $term->name );
		$sanitized_term_name = wp_strip_all_tags( $sanitized_term_name );
		if ( $sanitized_term_name[0] === '"' && $sanitized_term_name[ ( strlen( $sanitized_term_name) - 1 ) ] === '"' ) {
			    $sanitized_term_name = substr( $sanitized_term_name, 1, -1 );
		}

		if ( $term !== null ) {
			$this->data[ $attribute ] = [
				'@type' => 'Organization',
				'name'  => $tag_stripped_term_name,
			];
		}

File: classes/woocommerce-schema.php [hunk 2]
```diff
@@ -300,12 +300,20 @@ private function add_manufacturer( $product ) {
...
 	 */
 	private function add_organization_for_attribute( $attribute, $product, $taxonomy ) {
-		$term = $this->get_primary_term_or_first_term( $taxonomy, $product->get_id() );
+		$term     = $this->get_primary_term_or_first_term( $taxonomy, $product->get_id() );
+		$termname = $term->name;
+
+		// Escape the string so that harmful input data will not be executed.
+		$escaped_term_name      = WPSEO_Utils::format_json_encode( $termname );
+		$tag_stripped_term_name = wp_strip_all_tags( $escaped_term_name );
+		if ( $tag_stripped_term_name[0] === '"' && $tag_stripped_term_name[ ( strlen( $tag_stripped_term_name ) - 1 ) ] === '"' ) {
```
Comment (karlijnbok) [inline classes/woocommerce-schema.php:309]: Possibly add comment stating something like: "// Remove possible double quotes."

**Comment by karlijnbok** [comment]
Discussed CR with @JesserH, already applied changes.

**Comment by karlijnbok** [comment]
Double quotes are encoded two times with `WPSEO_Utils::format_json_encode`: 
- once in `add_organization_for_attribute` (in `woocommerce-schema.php`), and
- once in `schema_tag`(in `class-wpseo-utils.php`)

Changing the name in the Database to: `"Gimli"` changes the brand name in the schema output to: 
Screenshot 2020-04-03 at 13 59 40

Echoing `$term->name`, `$escaped_term_name` and `$stripped_term_name` results in:
Screenshot 2020-04-03 at 14 14 40

Might be a good idea to look in Free to see how the schema output is ensured and escaped there.

**Comment by JesserH** [comment]
@karlijnbok Updated the logic so the json encode is not triggered in the function itself, which makes every input with a tag (for instance: "</script><script>alert(1)</script><script>") appear on the product page itself but does not render anything in the schema. I think this solves the problem, what are your thoughts?
The entering of "Gimli" (with quotes) in the database conforming the test instructions now renders "\"Gimli\"", which i believe is wished behaviour

**Comment by karlijnbok** [comment]
CR & Acceptance: 👍

**Pull request merged**